### PR TITLE
Add new Primitive element to the AST

### LINKED ISSUE #56: Add AST support for primitive functions
To support functions like `print` which interact with the outside world we need to have a way to distinguish them from the regular functions we have in the language.

One way to do this is to add a new node to the AST for primitive/built-in functions. Such that calls to these functions trigger a particular method inside `Machine` (See https://github.com/christianpoveda/pijama/issues/55).

One prerequisite to fully implement this is that the AST must support arbitrary function calls (https://github.com/christianpoveda/pijama/issues/54) to allow the new primitive nodes in the function position. So we won't take care of the parsing part in this issue.

After https://github.com/christianpoveda/pijama/issues/55 is solved. We need to
- [ ] Add a new enum `Primitive` (or `BuiltIn`) with no variants for now.
    - [ ] Add a new `PrimitiveFn` (or `BuiltInFn`) variant to the `ast::Node` enum containing the above type.
    - [ ] Add a new `PrimitiveFn` (or `BuiltInFn`) variant to the `mir::Term` enum containing the above type.
    - [ ] Add a new `PrimitiveFn` (or `BuiltInFn`) variant to the `mir::Term` enum containing the above type.
- [ ] Add a new `mir::lower_primitive` (or `mir::lower_built_in`) method to transform the `Node::PrimitiveFn` variant to a `mir::Term::PrimitiveFn`.
- [ ] Update `ty::ty_check` to return `todo!()` when matching a `Term::PrimitiveFn`.
- [ ] Change the `lir::ctx::Context::remove_names` method to transform the `mir::Term::PrimitiveFn` variant to a `lir::Term::PrimitiveFn`.
- [ ] Add a new `Machine::step_primitive_app` method to handle terms with the pattern
    ```rust
    Term::Abs(box Term::PrimitiveFn(_), box t1)
    ```
    which returns `todo!()` for now.
- [ ] Dispatch the method above inside `Machine::step`.

### PR DESCRIPTION
**Pull request by pvdrz**
Fixes https://github.com/christianpoveda/pijama/issues/56.

@DarkDrek do you want to take a look?

### CODE REVIEW

File: src/mir/mod.rs [hunk 1]
```diff
@@ -43,6 +44,7 @@ impl<'a> Display for Term<'a> {
             Term::Let(name, t1, t2) => write!(f, "(let {} = {} in {})", name, t1, t2),
             Term::Seq(t1, t2) => write!(f, "{} ; {}", t1, t2),
             Term::Fix(t1) => write!(f, "(fix {})", t1),
+            Term::PrimFn(prim) => write!(f, "{},", prim),
```
Comment (DarkDrek) [inline src/mir/mod.rs]: Is the `,` wanted / a typo in the string?

File: src/ty/ty_check.rs [hunk 2]
```diff
@@ -155,6 +155,9 @@ impl<'a> Context<'a> {
                     return Err(TyError::ExpectedFn(ty));
                 }
             }
+            Term::PrimFn(prim) => match prim {
+                _ => panic!("There aren't any primitives yet!"),
```
Comment (DarkDrek) [inline src/ty/ty_check.rs]: Use `todo!` here too since we want to add them later?

File: src/machine/eval.rs [hunk 3]
```diff
@@ -117,6 +115,10 @@ impl Machine {
         // return the body
         (true, body)
     }
+    /// Evaluation step for application of primitive functions (prim arg)
+    fn step_primitive_app(&mut self, _prim: Primitive, _arg: Box<Term>) -> (bool, Term) {
+        todo!()
```
Comment (DarkDrek) [inline src/machine/eval.rs]: Just for completeness sake can you add `"There aren't any primitives yet!"` as argument to `todo!`?

File: src/mir/mod.rs (hunk 1 above)
Comment (pvdrz) [reply src/mir/mod.rs]: is a typo :P I'll fix it

File: src/lir/mod.rs [hunk 4]
```diff
@@ -17,6 +17,7 @@ pub enum Term {
     Cond(Box<Term>, Box<Term>, Box<Term>),
     Fix(Box<Term>),
     Hole,
+    PrimFn(Primitive),
```
Comment (DarkDrek) [inline src/lir/mod.rs:19]: Should we move this element at least one place upwards since `Hole` has a special meaning?

Comment (pvdrz) [reply src/lir/mod.rs:19]: You're right.

Update: I amended the last commit as this change didn't justify a new one :P

**Review by DarkDrek — APPROVED** [review]

**Pull request merged**